### PR TITLE
Week 5 walkthrough

### PR DESCRIPTION
**Pull request by billyporter**
I also did the optional challenge that is "allow support for optional attendees and return the time that works for the highest number of optional attendees"

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -250,25 +258,114 @@ public void noConflicts() {
...
+  }
+
+  @Test
+  public void onlyOptional() {
+    // No mandatory attendees, just two optional attendees with several gaps in their schedules. Those gaps should be identified and returned.
+
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false), Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false), Arrays.asList(PERSON_B)),
+        new Event("Event 5", TimeRange.fromStartEnd(TIME_0900AM, TimeRange.END_OF_DAY + 1, false), Arrays.asList(PERSON_C)));
```
Comment (NicholasCobbFromGoogle) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:346]: nit: I see for the time ranges for the end arguments you go between "TimeRange.END_OF_DAY + 1, inclusive=false" and "TimeRange.END_OF_DAY, inclusive=true. Is there a reason this is needed that I do not see? If not, then can we consistently use one or the other?

**Review by NicholasCobbFromGoogle — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 1 above)
Comment (billyporter) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:346]: Good catch. Removed the +1 and switched it to inclusive.

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 2]
```diff
@@ -30,4 +30,23 @@ public void testGreeting() {
 
     Assert.assertEquals("Hello Ada", greeting);
   }
+
```
Comment (mbehrlich) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java:33]: No action required, but just a note on a method like this and its tests. If the user submits a string where none of them are letters, like "^$@$*(@", then the method will return "Hello ". This is where we'd usually throw an error of some kind. If we did that, another test would make sure that an error was thrown with that input.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,160 @@
...
+        // convert bit to an int array of pairs of intervals
+        int[] openingsArray = masterOpenings.stream()
+                .filter(s -> (s == 0 || (!masterOpenings.get(s - 5)) || !masterOpenings.get(s + 5))).toArray();
+
+        // build collection
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+        int addStart, addEnd, start, end;
+
+        // Loop through array, if bigger than duration then add to output
+        for (int a = 0; a < openingsArray.length; a += 2) {
```
Comment (mbehrlich) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:68]: I'd recommend sticking with "i" for index.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,160 @@
...
+        return meetingTimes;
+    }
+
+    /*
+     * Function to convert a List of time ranges to a single bitset Can be an
+     * arbitrary length of time ranges
+     */
+    public BitSet timeRangeToBitSet(ArrayList<TimeRange> timeRanges, int sizeOfBitSet) {
+        BitSet condensedSet = new BitSet(sizeOfBitSet);
+        for (int c = 0; c < timeRanges.size(); c++) {
```
Comment (mbehrlich) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:90]: Recommend using i for index.

**Review by mbehrlich — APPROVED** [review]

**Pull request merged**